akka-remote/src/main/scala/akka/remote/artery/driver/FlowAndLossControl.scala

```diff
@@ -0,0 +1,246 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.remote.artery.driver
+
+import akka.NotUsed
+import akka.stream.impl.FixedSizeBuffer
+import akka.stream.scaladsl.BidiFlow
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.stream.stage._
+import akka.util.ByteString
+
+import scala.concurrent.duration._
+
+/**
+ * INTERNAL API
+ */
+private[remote] object FlowAndLossControl {
+  val WindowSize = 16
+  val BufferMask = WindowSize - 1
+  val AckWaterMark = WindowSize >> 1
+  val DATA: Byte = 0
+  val ACK: Byte = 1
+
+  val AckTimer = "AckTimer"
+  val AckPeriod = 10.millis
+
+  def apply(frameBuffer: FrameBuffer): BidiFlow[ByteString, Frame, Frame, ByteString, NotUsed] = {
+    BidiFlow.fromGraph(new FlowAndLossControlStage(frameBuffer))
+  }
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] final class FlowAndLossControlStage(sndFramePool: FrameBuffer) extends GraphStage[BidiShape[ByteString, Frame, Frame, ByteString]] {
```

Comment:
This class should be factored out into inbound and outbound buffers so that they encapsulate the necessary states for these directions and can be tested in isolation.